homeassistant/block_async_io.py

```diff
@@ -46,53 +49,139 @@ def _check_sleep_call_allowed(mapped_args: dict[str, Any]) -> bool:
+        strict_core=False,
+        skip_for_tests=False,
+    ),
+    BlockingCall(
+        original_func=glob.iglob,
+        module=glob,
+        function="iglob",
+        check_allowed=None,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=False,
+    ),
+    BlockingCall(
+        original_func=os.walk,
+        module=os,
+        function="walk",
+        check_allowed=None,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=False,
+    ),
+    BlockingCall(
+        original_func=os.listdir,
+        module=os,
+        function="listdir",
+        check_allowed=None,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=True,
+    ),
+    BlockingCall(
+        original_func=os.scandir,
+        module=os,
+        function="scandir",
+        check_allowed=None,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=True,
+    ),
+    BlockingCall(
+        original_func=builtins.open,
+        module=builtins,
+        function="open",
+        check_allowed=_check_file_allowed,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=True,
+    ),
+    BlockingCall(
+        original_func=importlib.import_module,
+        module=importlib,
+        function="import_module",
+        check_allowed=_check_import_call_allowed,
+        strict=False,
+        strict_core=False,
+        skip_for_tests=True,
+    ),
+)
+
+
+@dataclass(slots=True)
+class BlockedCalls:
+    """Class to track which calls are blocked."""
+
+    calls: set[BlockingCall]
+
+
+_BLOCKED_CALLS = BlockedCalls(set())
+
+
 def enable() -> None:
     """Enable the detection of blocking calls in the event loop."""
-    loop_thread_id = threading.get_ident()
-    # Prevent urllib3 and requests doing I/O in event loop
-    HTTPConnection.putrequest = protect_loop(  # type: ignore[method-assign]
-        HTTPConnection.putrequest, loop_thread_id=loop_thread_id
-    )
-
-    # Prevent sleeping in event loop.
-    time.sleep = protect_loop(
-        time.sleep,
-        check_allowed=_check_sleep_call_allowed,
-        loop_thread_id=loop_thread_id,
-    )
-
-    glob.glob = protect_loop(
-        glob.glob, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-    glob.iglob = protect_loop(
-        glob.iglob, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-    os.walk = protect_loop(
-        os.walk, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-
-    if not _IN_TESTS:
-        # Prevent files being opened inside the event loop
-        os.listdir = protect_loop(  # type: ignore[assignment]
-            os.listdir, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-        )
-        os.scandir = protect_loop(  # type: ignore[assignment]
-            os.scandir, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-        )
+    if _BLOCKED_CALLS.calls:
+        return
```

Comment:
I don't think this is ever expected to happen, maybe raise here instead of silently returning?

Reply:
changed it to raise.  It shouldn't ever need to be done twice